### ISSUE TITLE
Class based Sampler for Class Incremental/Continual Learning research

### ISSUE THREAD

**Issue by virajmavani** [pytorch/pytorch#20416]
## 🚀 Feature

A class based dataset sampler for class incremental and continual learning research.

## Motivation

For research in class incremental learning domain, we need datasets to be split up on the basis of classes so that after training on one subset of classes, others can be introduced and the model can be made to generalize on that too in an architectural fashion or using generalization techniques like Elastic Weight Consolidation. Pytorch is strong enough with dynamic graphs but currently it lacks a sampler which can return examples from specific classes.

## Pitch

I would like to add a sampler similar to others but which takes in the class labels as argument and returns examples from those classes. A random flag can be added to the arguments to randomize the order of samples in the subset. The implementation would be similar to that of SubsetRandomSampler in torch.utils.data but instead of indices, it would take in the class labels as argument.

## Alternatives

Please feel free to suggest any alternative implementations. :)

**Comment by SsnL** [comment]
This seems a perfect place to use a custom sampler class.

**Comment by virajmavani** [comment]
Shall I go ahead and start working on the implementation?

**Comment by dmgolembiowski** [comment]
This idea seems wonderful! Is there an appropriate LR algorithim to go along with the sampler class for this sort of undertaking? I have so many questions but I'm not sure which ones are the right ones to ask!

**Comment by soumith** [comment]
before anyone starts working on it -- what exactly is being proposed here?

I cant see how this can actually be implemented without enforcing constraints on the Dataset (and how / what constraints is not clear from the proposal)

**Comment by SsnL** [comment]
Sorry if I caused any confusion. My comment meant that this should be implemented on user side / or in a noncore library.

**Comment by virajmavani** [comment]
It is supposed to be implemented without placing any constraints on the Dataset. It will be a similar implementation as that of the subset classifier. Just instead of indices, it will use the class labels. @soumith @SsnL Still, if you think it is not needed in the source code, it's okay. It makes complete sense in implementing it as a custom sampler on user end. I was just proposing it so that everyone does not need to do it.

**Comment by soumith** [comment]
@virajmavani what is the proposed API?
